### PR TITLE
[android] build and ship libmono-profiler-aot.so

### LINKED ISSUE #56989: Support for Android "Profiled AOT" in .NET 6
This feature in Xamarin.Android is also known as "Startup Tracing":

https://devblogs.microsoft.com/xamarin/faster-android-startup-times-with-startup-tracing/

Yet another name for this might be "profile guided optimization".

For .NET 6, we'd like users be able to do something like the following in Release builds:

```xml
<RunAOTCompilation>true</RunAOTCompilation>
<UseProfiledAOT>true</UseProfiledAOT>
```

This would use a built-in profile that the Android workload provides as well as the .NET MAUI workload. Recording a custom profile for a developer's app is probably not something we need in .NET 6. However, the Android & .NET MAUI teams would need to be able to record profiles. .NET MAUI's startup path is still quite a bit in flux, so ideally we would be automate the creation of the built-in MAUI profile.

There are a few pieces missing from dotnet/runtime at the moment:

1. `libmono-profiler-aot.so`
2. `aprofutil`
3. `MonoAOTCompiler.AotProfilePath` only accepts 1 value:

https://github.com/dotnet/runtime/blob/ce3447c78fbddfe8485d6d7e527c1e8a7429d8f0/src/tasks/AotCompilerTask/MonoAOTCompiler.cs#L111-L114

For .NET 6, I think we can take a simple approach:

1. Build & ship `libmono-profiler-aot.so` in the Android runtime packs.
2. The Android workload can ship a `aprofutil` from mono/mono. We can emit warnings that mention recording custom AOT profiles is an experimental feature.
3. We can emit `profile-only,profile=foo.aprof` for `%(AotArguments)` in the Android MSBuild targets until there is a different way.

These would get us by for now.

In the long run, recording custom AOT profiles might be something that `dotnet-trace` / diagnostics should support.

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Context: https://github.com/xamarin/xamarin-android/pull/6171
Fixes: https://github.com/dotnet/runtime/issues/56989

In order for the Android workload to be able to record
`.aotprof`/`.aotprofile` files, we need `libmono-profiler-aot.so`
to be available. Down the road this feature could be provided by the
Mono diagnostics component, but will probably not happen in .NET 6.

These changes build `libmono-profiler-aot.so` for Android, and includes
it in the `Microsoft.NETCore.App.Runtime.Mono.android-*` runtime packs.
In the Android workload's MSBuild targets we exclude this native library
unless the app is configured to record an AOT profile.

I also included in `CMakeLists.txt`:

    target_compile_definitions(mono-profiler-aot PRIVATE -DMONO_DLL_EXPORT)

Otherwise, the AOT profiler cannot be loaded:

    08-12 16:01:39.817  3003  3003 I monodroid-assembly: Trying to load shared library '/data/app/com.microsoft.net6.helloandroid-4u8tNHoPAh4zSZMaf2FsnA==/lib/x86_64/libmono-profiler-aot.so'
    08-12 16:01:39.818  3003  3003 W monodroid: Looking for profiler init symbol 'mono_profiler_init_aot'? 0x0
    08-12 16:01:39.818  3003  3003 W monodroid: The 'aot' profiler wasn't found in the main executable nor could it be loaded from 'libmono-profiler-aot.so'.

With these changes, I can successfully record an AOT profile on Android:

    Reading from '127.0.0.1:9999'...
    Read 4096 bytes...
    ...
    Read 2671 bytes...
    Read total 72303 bytes...
    Summary:
            Modules:          8
            Types:          197
            Methods:        910
    Going to write the profile to 'custom.aprof'

When using the profile, I get improved startup times:

    08-12 16:56:33.940  1624  1874 I ActivityTaskManager: Displayed com.microsoft.net6.helloandroid/crc6490bfc84a0f5dff7a.MainActivity: +217ms

### CODE REVIEW

File: src/mono/mono/profiler/aot.c [hunk 1]
```diff
@@ -341,7 +341,7 @@ runtime_initialized (MonoProfiler *profiler)
 		start_helper_thread ();
 }
 
-MONO_API void
+MONO_API_EXPORT void
```
Comment (lambdageek) [inline src/mono/mono/profiler/aot.c:344]: Don't use `MONO_API_EXPORT` directly.

Set this in cmake:

```
    target_compile_definitions(mono-profiler-aot PRIVATE -DMONO_DLL_EXPORT)
```

**Review by akoeplinger — APPROVED** [review]

**Comment by jonathanpeppers** [comment]
Is there a build artifact I can double-check the contents of the `Microsoft.NETCore.App.Runtime.Mono.android-*` runtime packs?

I tried to find it, but there are 100+ of artifacts.

**Comment by lewing** [comment]
https://dev.azure.com/dnceng/public/_build/results?buildId=1296179&view=logs&j=9db4066d-6bf0-549a-7716-e181239d2ea7&t=ee7de8a6-b1ed-5700-3f81-b0654bf893a4&l=183

is https://github.com/dotnet/core-eng/issues/13691

**Comment by jonathanpeppers** [comment]
Closing this, as we've gone the route of not supporting custom profiles in .NET 6. In some future release, support for custom profiles with the `dotnet trace` infrastructure is probably the way to go.

To record a profile *now*, I've setup a repo:

https://github.com/jonathanpeppers/android-profiled-aot

We setup premade profiles for `dotnet new android` and `dotnet new maui` that are shipped in each workload.

**Pull request closed without merging**